File: lambda/teams_client.py
"""Microsoft Teams client for posting messages."""

import json
import urllib3
import logging
from typing import Dict, Any

logger = logging.getLogger(__name__)

# Global connection pool
http = urllib3.PoolManager()


def post_to_teams(event_data: Dict[str, Any], webhook_url: str) -> bool:
    """
    Post event data to Microsoft Teams Workflow.
    
    Args:
        event_data: Parsed event data for Teams Workflow template
        webhook_url: Teams Workflow webhook URL
    
    Returns:
        bool: True if posting succeeded, False otherwise
    """
    if not event_data:
        logger.error("Cannot post empty event data to Teams")
        return False
    
    if not webhook_url:
        logger.error("Teams webhook URL is required")
        return False
    
    try:
        # For Teams Workflows, we send the data directly
        # The workflow will use this data in the Adaptive Card template
        payload = event_data
        
        # Convert payload to JSON string
        payload_json = json.dumps(payload)
        
        # Prepare headers
        headers = {
            'Content-Type': 'application/json',
            'User-Agent': 'Bitbucket-Teams-Webhook/2.0'
        }
        
        # Extract title for logging
        title = event_data.get("title", "Unknown")
        
        logger.info(f"Posting event data to Teams: {title}")
        
        response = http.request(
            'POST',
            webhook_url,
            body=payload_json,
            headers=headers,
            timeout=10.0  # 10 second timeout
        )
        
        # Teams Workflows typically return 202 (Accepted) for successful posts
        if response.status in [200, 202]:
            logger.info("Successfully posted event data to Teams")
            return True
        else:
            # Log the error response for debugging (with status code and response body)
            error_body = response.data.decode('utf-8') if response.data else 'No response body'
            logger.error(f"Teams workflow returned status {response.status}: {error_body}")
            return False
            
    except Exception as e:
        logger.error(f"Error posting event data to Teams: {str(e)}")
        return False

File: src/webhook/teamsClient.ts
/**
 * Teams client for posting messages to Microsoft Teams
 * Posts formatted failure notifications to Teams webhook URL
 */

import * as https from 'https';
import { TeamsMessage } from '../types';
import { Logger } from '../logger';

/**
 * Post a message to Teams webhook URL
 * Handles HTTP status 200 and 202 as success
 * Logs errors for failed posts
 *
 * @param message The Teams message to post
 * @param webhookUrl The Teams webhook URL
 * @param logger The logger instance for error logging
 * @returns Promise<boolean> - true if post was successful, false otherwise
 */
export async function postToTeams(
  message: TeamsMessage,
  webhookUrl: string,
  logger: Logger
): Promise<boolean> {
  return new Promise((resolve) => {
    try {
      // Parse the webhook URL
      const url = new URL(webhookUrl);

      // Prepare the payload
      const payload = JSON.stringify({
        title: message.title,
        description: message.description,
        link: message.link,
        color: message.color,
      });

      // Prepare request options
      const options = {
        hostname: url.hostname,
        port: url.port || 443,
        path: url.pathname + url.search,
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Content-Length': Buffer.byteLength(payload),
        },
        timeout: 10000, // 10 second timeout
      };

      // Make the request
      const req = https.request(options, (res) => {
        let data = '';

        res.on('data', (chunk) => {
          data += chunk;
        });

        res.on('end', () => {
          // Handle success responses (200 or 202)
          if (res.statusCode === 200 || res.statusCode === 202) {
            logger.info('Successfully posted message to Teams', {
              statusCode: res.statusCode,
            });
            resolve(true);
          } else {
            // Log error for non-success responses
            logger.error('Teams API returned error', {
              statusCode: res.statusCode,
              response: data,
            });
            resolve(false);
          }
        });
      });

      // Handle request errors
      req.on('error', (error) => {
        logger.error('Failed to post message to Teams', {
          error: error.message,
        });
        resolve(false);
      });

      // Handle timeout
      req.on('timeout', () => {
        req.destroy();
        logger.error('Teams API request timed out', {
          timeout: 10000,
        });
        resolve(false);
      });

      // Send the payload
      req.write(payload);
      req.end();
    } catch (error) {
      logger.error('Error posting to Teams', {
        error: error instanceof Error ? error.message : String(error),
      });
      resolve(false);
    }
  });
}
